internal/storage/image.go

```diff
@@ -398,6 +412,145 @@ func (svc *imageService) ImageStatus(systemContext *types.SystemContext, nameOrI
 	return &result, nil
 }
 
+// nolint:staticcheck
+func (svc *imageService) IsRunningImageAllowed(ctx context.Context, systemContext *types.SystemContext, configImage, userSpecifiedImage string, auth *runtimeapi.AuthConfig) error {
+	if userSpecifiedImage == "" {
+		if systemContext.SignaturePolicyPath == "" {
+			log.Infof(ctx, "User specified image is empty and no policy applied, skipping check if image is allowed to run")
+			return nil
+		}
+
+		return errors.New("user specified image is empty")
+	}
+
+	log.Debugf(ctx, "Getting image status of config image: %s", configImage)
+	configImageStatus, err := svc.ImageStatus(systemContext, configImage)
+	if err != nil {
+		return fmt.Errorf("image status from config image: %w", err)
+	}
+
+	// Check the cache first
+	var policyContext *signature.PolicyContext
+	if cacheItem := svc.policyContextCache.Get(systemContext.SignaturePolicyPath); cacheItem != nil {
+		log.Debugf(ctx, "Using policy context from cache")
+		policyContext = cacheItem.Value()
+	} else {
+		log.Debugf(ctx, "Creating policy context")
+		policy, err := signature.DefaultPolicy(systemContext)
+		if err != nil {
+			return fmt.Errorf("get default policy: %w", err)
+		}
+
+		policyContext, err = signature.NewPolicyContext(policy)
+		if err != nil {
+			return fmt.Errorf("create policy context: %w", err)
+		}
+
+		log.Debugf(ctx, "Caching policy context")
+		svc.policyContextCache.Set(systemContext.SignaturePolicyPath, policyContext, ttlcache.DefaultTTL)
+	}
+
+	log.Debugf(ctx, "Parsing docker reference from %s", userSpecifiedImage)
+	imageRef, err := docker.ParseReference("//" + userSpecifiedImage)
+	if err != nil {
+		return err
+	}
+
+	// Check the auth cache first
+	authCacheItem := svc.authCache.Get(userSpecifiedImage)
```

Comment:
⚠️ Doesn’t this leak credentials across namespaces = tenants? Many tenants can use the same `example.com/vendor` repositories, each with their own credentials (and that might conceivably affect billing or usage metrics per customer).

---

(A reminder: https://github.com/cri-o/cri-o/pull/7046#discussion_r1240051558 is still an issue, and maybe needs solving first, to see if this cache is even needed.)

Reply:
I removed the auth cache for now, private registries can be supported later on.

> This checks whether the on-registry image is correctly signed. But nothing connects the on-registry image artifact with the already-locally-pulled config.Image, i.e. this does not verify anything about the image that is about to be run.

I tried something similar like we do in image pull, but I'm absolutely unsure if that's the right way: https://github.com/cri-o/cri-o/pull/7046/files#diff-73f26dcabd7e433970fc0612ff0068d2d8f7cd61dda790bcd603195197aaad97R465-R473